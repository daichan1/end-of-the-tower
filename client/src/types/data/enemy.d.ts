export type EnemyList = {
  stage1: EnemyType[]
  stage2: EnemyType[]
  stage3: EnemyType[]
  stage4: EnemyType[]
  stage5: EnemyType[]
  stage6: EnemyType[]
}

export type EnemyPattern = {
  pattern1: EnemyType[]
  pattern2: EnemyType[]
  pattern3: EnemyType[]
  pattern4: EnemyType[]
  pattern5: EnemyType[]
  pattern6: EnemyType[]
  pattern7: EnemyType[]
}
