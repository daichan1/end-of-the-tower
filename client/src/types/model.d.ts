export type EnemyType = {
  id: number
  name: string
  imageUrl: string
  hp: number
  attack: number
  defense: number
}

export type PlayerType = {
  name: string
  imageUrl: string
  hp: number
  attack: number
  defense: number
  energy: number
}

export type CardType = {
  name: string
  description: string
  imageUrl: string
  cost: number
  cardType: string
  attack: number
  defense: number
}
