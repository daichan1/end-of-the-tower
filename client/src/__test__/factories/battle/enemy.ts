import { configureStore, EnhancedStore } from "@reduxjs/toolkit"
import { EnemyType } from "../../../types/model"
import fightEnemiesReducer from '../../../redux/slice/fightEnemiesSlice'
import choiceEnemyReducer from '../../../redux/slice/choiceEnemySlice'

export type PreloadedState = {
  fightEnemies: EnemyType[]
  choiceEnemy: number
}

// テストデータ
const initialState: PreloadedState = {
  fightEnemies: [
    {
      id: 1,
      name: "スライム",
      imageUrl: "none",
      hp: 20,
      maxHp: 20,
      attack: 4,
      defense: 0,
      damage: 0,
      floorNumber: 1,
      isBoss: false
    }
  ],
  choiceEnemy: 0
}

// enemy component用のテスト用のstore
export const enemyStore = (preloadedState: PreloadedState = initialState): EnhancedStore => {
  return configureStore(
    {
      reducer: {
        fightEnemies: fightEnemiesReducer,
        choiceEnemy: choiceEnemyReducer
      },
      preloadedState
    }
  )
}
