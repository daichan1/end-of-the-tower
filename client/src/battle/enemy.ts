import { EnemyType } from '../types/model/index'

export const isExistEnemy = (enemies: EnemyType[]): boolean => {
  return enemies.length > 0 ? true : false
}

export const setDamage = (enemy: EnemyType, damage: number): void => {
  enemy.damage = damage
}
