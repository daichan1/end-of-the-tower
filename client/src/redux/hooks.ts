import { useSelector, useDispatch, TypedUseSelectorHook } from 'react-redux'
import type { RootStore, AppDispatch } from './store'

// 型付きのuseSelectorを事前に定義
export const useAppSelector: TypedUseSelectorHook<RootStore> = useSelector
// 型付きのuseDispatchを事前に定義
export const useAppDispatch = () => useDispatch<AppDispatch>()
