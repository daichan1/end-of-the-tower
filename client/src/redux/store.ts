import { configureStore } from '@reduxjs/toolkit'
import playerReducer from './slice/playerSlice'
import enemiesReducer from './slice/enemiesSlice'
import cardsReducer from './slice/cardsSlice'
import fightEnemiesReducer from './slice/fightEnemiesSlice'
import gameTitleReducer from './slice/gameTitleSlice'
import rootSelectReducer from './slice/rootSelectSlice'
import battleReducer from './slice/battleSlice'
import turnReducer from './slice/turnSlice'

export const store = configureStore({
  reducer: {
    player: playerReducer,
    enemies: enemiesReducer,
    cards: cardsReducer,
    fightEnemies: fightEnemiesReducer,
    gameTitle: gameTitleReducer,
    rootSelect: rootSelectReducer,
    battle: battleReducer,
    turn: turnReducer
  }
})

export type RootStore = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
