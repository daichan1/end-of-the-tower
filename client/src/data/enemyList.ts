import { EnemyType } from '../types/model/index'
import { EnemyList } from '../types/data/enemy'

export const createEnemyList = (enemies: EnemyType[]): EnemyList => {
  const enemyList: EnemyList = {
    slime: [],
    gargoyle: [],
    slimeAndGargoyle: []
  }
  enemyList.slime = enemyList.slime.concat(searchEnemy(enemies, ["スライム"]))
  enemyList.gargoyle = enemyList.gargoyle.concat(searchEnemy(enemies, ["ガーゴイル"]))
  enemyList.slimeAndGargoyle = enemyList.slimeAndGargoyle.concat(searchEnemy(enemies, ["スライム", "ガーゴイル"]))
  return enemyList
}

const searchEnemy = (enemies: EnemyType[], searchTextList: Array<string>): EnemyType[] => {
  let enemyList: EnemyType[] = []
  searchTextList.forEach((text) => {
    enemyList = enemyList.concat(enemies.filter((enemy) => enemy.name === text))
  })
  return enemyList
}
