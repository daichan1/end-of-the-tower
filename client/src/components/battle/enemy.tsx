import { styled } from '@mui/material/styles'
import ShieldIcon from '@mui/icons-material/Shield'
import ArrowDownwardIcon from '@mui/icons-material/ArrowDownward'
import Typography from '@mui/material/Typography'
import LinearProgress from '@mui/material/LinearProgress'
import { useAppSelector, useAppDispatch } from '../../redux/hooks'
import { EnemyType } from '../../types/model/index'
import { ChoiceEnemy } from '../../types/battle/index'
import { resetDamage } from '../../redux/slice/fightEnemiesSlice'
import { resetPlayerDamage } from '../../redux/slice/playerDamageSlice'
import { setChoiceEnemyNumber } from '../../redux/slice/choiceEnemySlice'
import uuid from '../../common/uuid'
import { hpAdjustment } from '../../common/battle'
import enemyImg from '../../images/enemy.png'

type Props = {
  enemy: EnemyType
  index: number
}

const CustomLinearProgress = styled(LinearProgress)({
  width: 100,
  margin: "auto"
})

const Enemy = (props: Props): JSX.Element => {
  const { enemy, index } = props
  const choiceEnemyNumber = useAppSelector((state) => state.choiceEnemy)
  const dispatch = useAppDispatch()

  const DisplayChoiceEnemy = (props: ChoiceEnemy): JSX.Element => {
    const { enemyNumber } = props
    if (choiceEnemyNumber === enemyNumber) {
      return <div><ArrowDownwardIcon /></div>
    } else {
      return <div></div>
    }
  }

  const enemyImageClick = (num: number): void => {
    dispatch(setChoiceEnemyNumber(num))
    dispatch(resetDamage())
    dispatch(resetPlayerDamage())
  }

  return (
    <div>
      <DisplayChoiceEnemy enemyNumber={index} />
      <img src={enemyImg} alt={enemy.name} className='enemy-img' onClick={() => enemyImageClick(index)} />
      <span className='damage' key={uuid()}>{enemy.damage < 0 ? "" : enemy.damage}</span>
      <CustomLinearProgress variant="determinate" value={hpAdjustment(enemy.hp, enemy.maxHp, 0)}/>
      <Typography variant="subtitle1" component="div">
        {enemy.hp}/{enemy.maxHp}
      </Typography>
      <div>
        <ShieldIcon />
        <span>{enemy.defense}</span>
      </div>
    </div>
  )
}

export default Enemy
