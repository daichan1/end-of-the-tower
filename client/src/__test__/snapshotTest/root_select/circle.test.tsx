import renderer from 'react-test-renderer'
import { useAppSelector, useAppDispatch } from '../../../redux/hooks'
import { PlayerType } from '../../../types/model'
import { EnemyList } from '../../../types/data/enemy'
import Circle from '../../../components/root_select/circle'

jest.mock('../../../redux/hooks')
const useAppSelectorMock = useAppSelector as jest.Mock
const useAppDispatchMock = useAppDispatch as jest.Mock

describe('Circle Component', () => {
  const player: PlayerType = {
    name: "アタッカー",
    imageUrl: "none",
    hp: 80,
    maxHp: 80,
    attack: 0,
    defense: 0,
    energy: 3,
    stage: 0,
    deck: [],
    nameplate: [],
    cemetery: []
  }
  const enemyList: EnemyList = {
    stage1: [],
    stage2: [],
    stage3: [],
    stage4: [],
    stage5: [],
    stage6: []
  }
  beforeEach(() => {
    useAppSelectorMock.mockReturnValueOnce(player)
    useAppSelectorMock.mockReturnValueOnce(enemyList)
    useAppDispatchMock.mockReturnValue(jest.fn())
  })
  afterEach(() => {
    jest.resetAllMocks()
  })

  it("display not choice circle", () => {
    const component = renderer.create(
      <Circle left={10} top={510} rootNumber={1} />
    )
    .toJSON()

    expect(component).toMatchSnapshot()
  })

  it("display choice circle", () => {
    const component = renderer.create(
      <Circle left={10} top={510} rootNumber={0} />
    )
    .toJSON()

    expect(component).toMatchSnapshot()
  })

  it("display clear circle", () => {
    player.stage = 1
    const component = renderer.create(
      <Circle left={10} top={510} rootNumber={0} />
    )
    .toJSON()

    expect(component).toMatchSnapshot()
  })
})
