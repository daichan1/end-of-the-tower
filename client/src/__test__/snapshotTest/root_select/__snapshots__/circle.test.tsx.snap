// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle Component display choice circle 1`] = `
<div>
  <button
    className="circle
        next-stage
        
        "
    onClick={[Function]}
    style={
      Object {
        "left": "10%",
        "top": 510,
      }
    }
  />
</div>
`;

exports[`Circle Component display clear circle 1`] = `
<div>
  <button
    className="circle
        
        clear
        "
    onClick={[Function]}
    style={
      Object {
        "left": "10%",
        "top": 510,
      }
    }
  />
</div>
`;

exports[`Circle Component display not choice circle 1`] = `
<div>
  <button
    className="circle
        
        
        "
    onClick={[Function]}
    style={
      Object {
        "left": "10%",
        "top": 510,
      }
    }
  />
</div>
`;
